### PR TITLE
Automatic Sync initializing bugfix (Based on 1.7.2 patch PR)

### LINKED ISSUE #254: Fix Time Sync
With the changes to the coop framework, time control is no longer syncing. Investigate and fix this issue.

### PR DESCRIPTION
**Pull request by WakooMan**
## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] All new classes have class-level documentation comments, if there are any at all
- [x] Tests for the changes have been added (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

## What is the current behaviour?

- TimeControl sync does not work.
- CoopManaged child classes were not setup automatically, because constructors weren't called at deserialization.

## What is the new behaviour?

- Deserialized objects are now registered by CoopFramework, when they are loaded.
- TimeControl sync now works and TimeCommands written for testing.
## Resolves

- TimeControl bug
- Sync Automatic Setup bugfix

## Other information

### CODE REVIEW

File: source/CoopFramework/ObjectLifetimeObserver.cs [hunk 1]
```diff
@@ -61,14 +64,46 @@ public bool PatchConstruction()
...
+        ///     in the hierarchy, that has this attribute./>.
+        /// </summary>
+        private void PatchLoadInitializationCallbacks()
+        {
+            Type current = typeof(T);
+            m_LoadInitializationPatch = new List<MethodPatch<ObjectLifetimeObserver<T>>>();
+            while (current != typeof(object))
+            {
+                    var patch = new MethodPatch<ObjectLifetimeObserver<T>>(current);
+                    var methods = current.GetMethods(BindingFlags.Instance | BindingFlags.NonPublic | BindingFlags.Public).Where(m => m.GetCustomAttributes(typeof(LoadInitializationCallback), false).Any()).ToList();
```
Comment (garrettluskey) [inline source/CoopFramework/ObjectLifetimeObserver.cs]: This should also get parent class methods, you should be able to remove the while loop and `current = current.BaseType`

**Review by garrettluskey — CHANGES_REQUESTED** [review]

File: source/CoopFramework/ObjectLifetimeObserver.cs (hunk 1 above)
Comment (WakooMan) [reply source/CoopFramework/ObjectLifetimeObserver.cs]: Removed the while loop, but changed the whole thing. Please take a look at it. Also Movement sync does not work somehow, but I don't know if it was working before.

File: source/CoopFramework/CoopFramework.cs [hunk 2]
```diff
@@ -38,6 +42,16 @@ public static void InitPatches([CanBeNull] IObjectManager objectManager, Func<bo
             }
 
             var loadedAssemblies = AppDomain.CurrentDomain.GetAssemblies();
+            loadedAssemblies.ToList().ForEach(a =>
```
Comment (garrettluskey) [inline source/CoopFramework/CoopFramework.cs]: Split these into helper functions. Too many anonymous methods

File: source/CoopFramework/ObjectLifetimeObserver.cs [hunk 3]
```diff
@@ -58,17 +62,43 @@ public void AfterUnregisterObject(object removedObject)
...
         }
+        /// <summary>
+        ///     Patches the first method with <see cref="LoadInitializationCallback"/> LoadInitializationCallback attribute, but if none is found, then we check 
+        ///     if T is subclass of MBObjectBase and patch it's first method with the given attribute.. 
+        ///     This method is needed because constructors are not called at some classes, when loading a saved game, but for these classes there is a method
+        ///     in the hierarchy, that has this attribute./>.
+        /// </summary>
+        private void PatchLoadInitializationCallbacks()
+        {
+            Type type = CoopFramework.LoadInitializationCallbacks.Keys.ToList().Find(t => typeof(T) == t);
```
Comment (garrettluskey) [inline source/CoopFramework/ObjectLifetimeObserver.cs]: Try this, you can remove one of the null checks then as well assuming it works as intended

```suggestion
            Type type = CoopFramework.LoadInitializationCallbacks.Keys.ToList().Find(t => t.IsAssignableFrom(typeof(T));
```

**Review by garrettluskey — CHANGES_REQUESTED** [review]

File: source/CoopFramework/ObjectLifetimeObserver.cs (hunk 3 above)
Comment (WakooMan) [reply source/CoopFramework/ObjectLifetimeObserver.cs]: Done

File: source/CoopFramework/CoopFramework.cs (hunk 2 above)
Comment (WakooMan) [reply source/CoopFramework/CoopFramework.cs]: foreach cycles are no longer anonymous methods and added one local variable for BindingFlags, so it is easier to read in my opinion.

**Review by garrettluskey — APPROVED** [review]

**Pull request merged**